diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
index a451bc4be7..1e812d092a 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
@@ -404,22 +404,10 @@ void CBaseRenderer::SetViewMode(int viewMode)
   CDisplaySettings::GetInstance().SetVerticalShift(0.0f);
   CDisplaySettings::GetInstance().SetNonLinearStretched(false);
 
-  if (m_videoSettings.m_ViewMode == ViewModeZoom ||
-       (is43 && CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_STRETCH43) == ViewModeZoom))
-  { // zoom image so no black bars
+  if (m_videoSettings.m_ViewMode == ViewModeZoom)
+  { // miss use as 200%
     CDisplaySettings::GetInstance().SetPixelRatio(1.0);
-    // calculate the desired output ratio
-    float outputFrameRatio = sourceFrameRatio * CDisplaySettings::GetInstance().GetPixelRatio() / info.fPixelRatio;
-    // now calculate the correct zoom amount.  First zoom to full height.
-    float newHeight = screenHeight;
-    float newWidth = newHeight * outputFrameRatio;
-    CDisplaySettings::GetInstance().SetZoomAmount(newWidth / screenWidth);
-    if (newWidth < screenWidth)
-    { // zoom to full width
-      newWidth = screenWidth;
-      newHeight = newWidth / outputFrameRatio;
-      CDisplaySettings::GetInstance().SetZoomAmount(newHeight / screenHeight);
-    }
+    CDisplaySettings::GetInstance().SetZoomAmount(2.0);
   }
   else if (m_videoSettings.m_ViewMode == ViewModeStretch4x3)
   { // stretch image to 4:3 ratio
@@ -451,19 +439,9 @@ void CBaseRenderer::SetViewMode(int viewMode)
     CDisplaySettings::GetInstance().SetNonLinearStretched(nonlin);
   }
   else  if (m_videoSettings.m_ViewMode == ViewModeOriginal)
-  { // zoom image so that the height is the original size
+  { // miss use as 1/3
     CDisplaySettings::GetInstance().SetPixelRatio(1.0);
-    // get the size of the media file
-    // calculate the desired output ratio
-    float outputFrameRatio = sourceFrameRatio * CDisplaySettings::GetInstance().GetPixelRatio() / info.fPixelRatio;
-    // now calculate the correct zoom amount.  First zoom to full width.
-    float newHeight = screenWidth / outputFrameRatio;
-    if (newHeight > screenHeight)
-    { // zoom to full height
-      newHeight = screenHeight;
-    }
-    // now work out the zoom amount so that no zoom is done
-    CDisplaySettings::GetInstance().SetZoomAmount(m_sourceHeight / newHeight);
+    CDisplaySettings::GetInstance().SetZoomAmount(0.333333);
   }
   else if (m_videoSettings.m_ViewMode == ViewModeCustom)
   {
diff --git a/xbmc/video/ViewModeSettings.cpp b/xbmc/video/ViewModeSettings.cpp
index 6c8a91976e..03efa9e3a9 100644
--- a/xbmc/video/ViewModeSettings.cpp
+++ b/xbmc/video/ViewModeSettings.cpp
@@ -27,15 +27,15 @@ struct ViewModeProperties
 static const ViewModeProperties viewModes[] =
 {
   { 630,   ViewModeNormal },
-  { 631,   ViewModeZoom },
-  { 39008, ViewModeZoom120Width },
-  { 39009, ViewModeZoom110Width },
-  { 632,   ViewModeStretch4x3 },
-  { 633,   ViewModeWideZoom },
-  { 634,   ViewModeStretch16x9 },
+  { 631,   ViewModeZoom, HIDE_ITEM },
+  { 39008, ViewModeZoom120Width, HIDE_ITEM, HIDE_ITEM },
+  { 39009, ViewModeZoom110Width, HIDE_ITEM, HIDE_ITEM },
+  { 632,   ViewModeStretch4x3, HIDE_ITEM, HIDE_ITEM },
+  { 633,   ViewModeWideZoom, HIDE_ITEM, HIDE_ITEM },
+  { 634,   ViewModeStretch16x9, HIDE_ITEM, HIDE_ITEM },
   { 644,   ViewModeStretch16x9Nonlin, HIDE_ITEM, HIDE_ITEM },
   { 635,   ViewModeOriginal },
-  { 636,   ViewModeCustom, HIDE_ITEM }
+  { 636,   ViewModeCustom }
 };
 
 #define NUMBER_OF_VIEW_MODES (sizeof(viewModes) / sizeof(viewModes[0]))
