commit 083e706f6a948949c1a1b4d68262496bd6ad42eb
Author: Peter Tuschy <sky42+git@mailbox.org>
Date:   Mon Jan 13 15:44:11 2025 +0100

    ZoomHack: Zoom120Width -> 50%

diff --git a/addons/resource.language.en_gb/resources/strings.po b/addons/resource.language.en_gb/resources/strings.po
index e7817e5dbc..6edb29aa21 100644
--- a/addons/resource.language.en_gb/resources/strings.po
+++ b/addons/resource.language.en_gb/resources/strings.po
@@ -23118,7 +23118,7 @@ msgstr ""
 
 #: xbmc/video/dialogs/GUIDialogVideoSettings.cpp
 msgctxt "#39008"
-msgid "Zoom - 120% width"
+msgid "Zoom - 50%"
 msgstr ""
 
 #: xbmc/video/dialogs/GUIDialogVideoSettings.cpp
diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
index a451bc4be7..ed4124c4b6 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
@@ -472,11 +472,10 @@ void CBaseRenderer::SetViewMode(int viewMode)
     CDisplaySettings::GetInstance().SetNonLinearStretched(m_videoSettings.m_CustomNonLinStretch);
     CDisplaySettings::GetInstance().SetVerticalShift(m_videoSettings.m_CustomVerticalShift);
   }
-  else if (m_videoSettings.m_ViewMode == ViewModeZoom120Width)
+  else if (m_videoSettings.m_ViewMode == ViewModeZoom50)
   {
-    float fitHeightZoom = sourceFrameRatio * screenHeight / (info.fPixelRatio * screenWidth);
-    CDisplaySettings::GetInstance().SetPixelRatio(1.0f);
-    CDisplaySettings::GetInstance().SetZoomAmount(fitHeightZoom < 1.0f ? 1.0f : (fitHeightZoom > 1.2f ? 1.2f : fitHeightZoom));
+    CDisplaySettings::GetInstance().SetPixelRatio(1.0);
+    CDisplaySettings::GetInstance().SetZoomAmount(0.5);
   }
   else if (m_videoSettings.m_ViewMode == ViewModeZoom110Width)
   {
diff --git a/xbmc/cores/VideoSettings.h b/xbmc/cores/VideoSettings.h
index a7135da5d8..d544c1bcc0 100644
--- a/xbmc/cores/VideoSettings.h
+++ b/xbmc/cores/VideoSettings.h
@@ -187,7 +187,7 @@ enum ViewMode
   ViewModeOriginal,
   ViewModeCustom,
   ViewModeStretch16x9Nonlin,
-  ViewModeZoom120Width,
+  ViewModeZoom50,
   ViewModeZoom110Width
 };
 
diff --git a/xbmc/interfaces/json-rpc/PlayerOperations.cpp b/xbmc/interfaces/json-rpc/PlayerOperations.cpp
index 5029549541..6d7541cf42 100644
--- a/xbmc/interfaces/json-rpc/PlayerOperations.cpp
+++ b/xbmc/interfaces/json-rpc/PlayerOperations.cpp
@@ -648,7 +648,7 @@ std::map<std::string, ViewMode> viewModes =
   {"stretch16x9", ViewModeStretch16x9},
   {"original", ViewModeOriginal},
   {"stretch16x9nonlin", ViewModeStretch16x9Nonlin},
-  {"zoom120width", ViewModeZoom120Width},
+  {"zoom50", ViewModeZoom50},
   {"zoom110width", ViewModeZoom110Width}
 };
 
diff --git a/xbmc/interfaces/json-rpc/schema/types.json b/xbmc/interfaces/json-rpc/schema/types.json
index 4c67f7c35e..2c020c6355 100644
--- a/xbmc/interfaces/json-rpc/schema/types.json
+++ b/xbmc/interfaces/json-rpc/schema/types.json
@@ -527,7 +527,7 @@
       "stretch16x9",
       "original",
       "stretch16x9nonlin",
-      "zoom120width",
+      "zoom50",
       "zoom110width"
     ]
   },
diff --git a/xbmc/video/ViewModeSettings.cpp b/xbmc/video/ViewModeSettings.cpp
index 6c8a91976e..f728f3d022 100644
--- a/xbmc/video/ViewModeSettings.cpp
+++ b/xbmc/video/ViewModeSettings.cpp
@@ -28,7 +28,7 @@ static const ViewModeProperties viewModes[] =
 {
   { 630,   ViewModeNormal },
   { 631,   ViewModeZoom },
-  { 39008, ViewModeZoom120Width },
+  { 39008, ViewModeZoom50 },
   { 39009, ViewModeZoom110Width },
   { 632,   ViewModeStretch4x3 },
   { 633,   ViewModeWideZoom },

commit 78bc0284d929581e9e300296b51efeb22177b7c6
Author: Peter Tuschy <sky42+git@mailbox.org>
Date:   Mon Jan 13 15:52:44 2025 +0100

    ZoomHack: Zoom110Width -> 33%

diff --git a/addons/resource.language.en_gb/resources/strings.po b/addons/resource.language.en_gb/resources/strings.po
index 6edb29aa21..fe749f937b 100644
--- a/addons/resource.language.en_gb/resources/strings.po
+++ b/addons/resource.language.en_gb/resources/strings.po
@@ -23123,7 +23123,7 @@ msgstr ""
 
 #: xbmc/video/dialogs/GUIDialogVideoSettings.cpp
 msgctxt "#39009"
-msgid "Zoom - 110% width"
+msgid "Zoom - 33%"
 msgstr ""
 
 #: view/GUIViewState.cpp
diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
index ed4124c4b6..3c09cddc6b 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
@@ -378,7 +378,7 @@ EShaderFormat CBaseRenderer::GetShaderFormat()
 
 void CBaseRenderer::SetViewMode(int viewMode)
 {
-  if (viewMode < ViewModeNormal || viewMode > ViewModeZoom110Width)
+  if (viewMode < ViewModeNormal || viewMode > ViewModeZoom33)
     viewMode = ViewModeNormal;
 
   m_videoSettings.m_ViewMode = viewMode;
@@ -477,11 +477,10 @@ void CBaseRenderer::SetViewMode(int viewMode)
     CDisplaySettings::GetInstance().SetPixelRatio(1.0);
     CDisplaySettings::GetInstance().SetZoomAmount(0.5);
   }
-  else if (m_videoSettings.m_ViewMode == ViewModeZoom110Width)
+  else if (m_videoSettings.m_ViewMode == ViewModeZoom33)
   {
-    float fitHeightZoom = sourceFrameRatio * screenHeight / (info.fPixelRatio * screenWidth);
-    CDisplaySettings::GetInstance().SetPixelRatio(1.0f);
-    CDisplaySettings::GetInstance().SetZoomAmount(fitHeightZoom < 1.0f ? 1.0f : (fitHeightZoom > 1.1f ? 1.1f : fitHeightZoom));
+    CDisplaySettings::GetInstance().SetPixelRatio(1.0);
+    CDisplaySettings::GetInstance().SetZoomAmount(0.333333);
   }
   else // if (CMediaSettings::GetInstance().GetCurrentVideoSettings().m_ViewMode == ViewModeNormal)
   {
diff --git a/xbmc/cores/VideoSettings.h b/xbmc/cores/VideoSettings.h
index d544c1bcc0..95c89abc36 100644
--- a/xbmc/cores/VideoSettings.h
+++ b/xbmc/cores/VideoSettings.h
@@ -188,7 +188,7 @@ enum ViewMode
   ViewModeCustom,
   ViewModeStretch16x9Nonlin,
   ViewModeZoom50,
-  ViewModeZoom110Width
+  ViewModeZoom33
 };
 
 class CVideoSettings
diff --git a/xbmc/interfaces/json-rpc/PlayerOperations.cpp b/xbmc/interfaces/json-rpc/PlayerOperations.cpp
index 6d7541cf42..4b50331e29 100644
--- a/xbmc/interfaces/json-rpc/PlayerOperations.cpp
+++ b/xbmc/interfaces/json-rpc/PlayerOperations.cpp
@@ -649,7 +649,7 @@ std::map<std::string, ViewMode> viewModes =
   {"original", ViewModeOriginal},
   {"stretch16x9nonlin", ViewModeStretch16x9Nonlin},
   {"zoom50", ViewModeZoom50},
-  {"zoom110width", ViewModeZoom110Width}
+  {"zoom110width", ViewModeZoom33}
 };
 
 std::string GetStringFromViewMode(ViewMode viewMode)
diff --git a/xbmc/settings/MediaSettings.cpp b/xbmc/settings/MediaSettings.cpp
index 09aad05861..be870164b7 100644
--- a/xbmc/settings/MediaSettings.cpp
+++ b/xbmc/settings/MediaSettings.cpp
@@ -85,7 +85,7 @@ bool CMediaSettings::Load(const TiXmlNode *settings)
       scalingMethod = (int)VS_SCALINGMETHOD_LINEAR;
     m_defaultVideoSettings.m_ScalingMethod = (ESCALINGMETHOD)scalingMethod;
 
-    XMLUtils::GetInt(pElement, "viewmode", m_defaultVideoSettings.m_ViewMode, ViewModeNormal, ViewModeZoom110Width);
+    XMLUtils::GetInt(pElement, "viewmode", m_defaultVideoSettings.m_ViewMode, ViewModeNormal, ViewModeZoom33);
     if (!XMLUtils::GetFloat(pElement, "zoomamount", m_defaultVideoSettings.m_CustomZoomAmount, 0.5f, 2.0f))
       m_defaultVideoSettings.m_CustomZoomAmount = 1.0f;
     if (!XMLUtils::GetFloat(pElement, "pixelratio", m_defaultVideoSettings.m_CustomPixelRatio, 0.5f, 2.0f))
diff --git a/xbmc/video/ViewModeSettings.cpp b/xbmc/video/ViewModeSettings.cpp
index f728f3d022..0d88ef1539 100644
--- a/xbmc/video/ViewModeSettings.cpp
+++ b/xbmc/video/ViewModeSettings.cpp
@@ -29,7 +29,7 @@ static const ViewModeProperties viewModes[] =
   { 630,   ViewModeNormal },
   { 631,   ViewModeZoom },
   { 39008, ViewModeZoom50 },
-  { 39009, ViewModeZoom110Width },
+  { 39009, ViewModeZoom33 },
   { 632,   ViewModeStretch4x3 },
   { 633,   ViewModeWideZoom },
   { 634,   ViewModeStretch16x9 },

commit d2e5b514d63e20abfd5c61372bf44f0fce1bfa16
Author: Peter Tuschy <sky42+git@mailbox.org>
Date:   Mon Jan 13 16:01:49 2025 +0100

    ZoomHack: Zoom -> 200%

diff --git a/addons/resource.language.en_gb/resources/strings.po b/addons/resource.language.en_gb/resources/strings.po
index fe749f937b..2431e2d8fa 100644
--- a/addons/resource.language.en_gb/resources/strings.po
+++ b/addons/resource.language.en_gb/resources/strings.po
@@ -2958,7 +2958,7 @@ msgstr ""
 #: system/settings/settings.xml
 #: xbmc/video/dialogs/GUIDialogVideoSettings.cpp
 msgctxt "#631"
-msgid "Zoom"
+msgid "Zoom - 200%"
 msgstr ""
 
 #. Label for stretching the displayed video to a 4:3 aspect ratio
diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
index 3c09cddc6b..25b4b05dd6 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
@@ -404,22 +404,10 @@ void CBaseRenderer::SetViewMode(int viewMode)
   CDisplaySettings::GetInstance().SetVerticalShift(0.0f);
   CDisplaySettings::GetInstance().SetNonLinearStretched(false);
 
-  if (m_videoSettings.m_ViewMode == ViewModeZoom ||
-       (is43 && CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_STRETCH43) == ViewModeZoom))
-  { // zoom image so no black bars
+  if (m_videoSettings.m_ViewMode == ViewModeZoom)
+  { // 200%
     CDisplaySettings::GetInstance().SetPixelRatio(1.0);
-    // calculate the desired output ratio
-    float outputFrameRatio = sourceFrameRatio * CDisplaySettings::GetInstance().GetPixelRatio() / info.fPixelRatio;
-    // now calculate the correct zoom amount.  First zoom to full height.
-    float newHeight = screenHeight;
-    float newWidth = newHeight * outputFrameRatio;
-    CDisplaySettings::GetInstance().SetZoomAmount(newWidth / screenWidth);
-    if (newWidth < screenWidth)
-    { // zoom to full width
-      newWidth = screenWidth;
-      newHeight = newWidth / outputFrameRatio;
-      CDisplaySettings::GetInstance().SetZoomAmount(newHeight / screenHeight);
-    }
+    CDisplaySettings::GetInstance().SetZoomAmount(2.0);
   }
   else if (m_videoSettings.m_ViewMode == ViewModeStretch4x3)
   { // stretch image to 4:3 ratio

commit e385f61d1245bd2b8fac3e9434a516fd61bc384f
Author: Peter Tuschy <sky42+git@mailbox.org>
Date:   Mon Jan 13 16:06:23 2025 +0100

    ZoomHack: show only modes for this hack

diff --git a/xbmc/video/ViewModeSettings.cpp b/xbmc/video/ViewModeSettings.cpp
index 0d88ef1539..3b2868489f 100644
--- a/xbmc/video/ViewModeSettings.cpp
+++ b/xbmc/video/ViewModeSettings.cpp
@@ -27,14 +27,14 @@ struct ViewModeProperties
 static const ViewModeProperties viewModes[] =
 {
   { 630,   ViewModeNormal },
-  { 631,   ViewModeZoom },
+  { 631,   ViewModeZoom, HIDE_ITEM },
   { 39008, ViewModeZoom50 },
   { 39009, ViewModeZoom33 },
-  { 632,   ViewModeStretch4x3 },
-  { 633,   ViewModeWideZoom },
-  { 634,   ViewModeStretch16x9 },
+  { 632,   ViewModeStretch4x3, HIDE_ITEM, HIDE_ITEM },
+  { 633,   ViewModeWideZoom, HIDE_ITEM, HIDE_ITEM },
+  { 634,   ViewModeStretch16x9, HIDE_ITEM, HIDE_ITEM },
   { 644,   ViewModeStretch16x9Nonlin, HIDE_ITEM, HIDE_ITEM },
-  { 635,   ViewModeOriginal },
+  { 635,   ViewModeOriginal, HIDE_ITEM },
   { 636,   ViewModeCustom, HIDE_ITEM }
 };
 
